test/boost/group0_voter_registry_test.cc

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "test/lib/scylla_test_case.hh"
+#include <seastar/core/coroutine.hh>
+#include <utility>
+
+#include "service/raft/group0_voter_registry.hh"
+#include "service/raft/raft_group0.hh"
+#include "service/topology_state_machine.hh"
+
+namespace {
+
+class server_info_accessor : public service::raft_server_info_accessor {
+    std::unordered_map<raft::server_id, service::replica_state> _state;
+
+public:
+    // Provide concrete implementations for all pure virtual functions
+    [[nodiscard]] const service::replica_state& find(raft::server_id id) const override {
+        auto it = _state.find(id);
+        if (it != _state.end()) {
+            return it->second;
+        }
+        throw std::runtime_error("server info missing!");
+    }
+
+    server_info_accessor& set_info(raft::server_id id, service::replica_state state) {
+        _state[id] = std::move(state);
+        return *this;
+    }
+};
+
+class raft_voter_client : public service::raft_voter_client {
+
+    std::unordered_set<raft::server_id> _voters;
+
+public:
+    // Provide concrete implementations for all pure virtual functions
+    future<> set_voters_status(const std::unordered_set<raft::server_id>& nodes, raft::can_vote can_vote, seastar::abort_source& as) override {
+        if (can_vote == raft::can_vote::yes) {
+            _voters.insert(nodes.begin(), nodes.end());
+        } else {
+            for (const auto& node : nodes) {
+                _voters.erase(node);
+            }
+        }
+        return seastar::make_ready_future<>();
+    }
+
+    const std::unordered_set<raft::server_id>& voters() const {
+        return _voters;
+    }
+};
+
+} // namespace
+
+
+SEASTAR_TEST_CASE(test_voter_registry_insert_single_node) {
+    server_info_accessor server_info_accessor;
```

Comment:
As for the unit tests (and the interface abstraction):
* I suppose it might be possible to make it work directly on the `raft_group0` instance (that is currently abstracted out) by using `do_with_cql_env()`, retrieving the `group0` and passing it to the `group0_voter_registry`
* however not sure yet how easy would be to get a good observability on the selected voters (but probably could get the configuration members and check their attributes for the selected voters)
* however abstracting that out appears to make the tests more readable (and really testing just the business logic of the voter registry)